docs/design/numeric_literals.md

```diff
@@ -0,0 +1,241 @@
+Primitive operators are available between numeric literals, and produce values
+with numeric literal types. For example, the type of `1 + 2` is the same as the
+type of `3`.
+
+Numeric types can provide conversions to support initialization from numeric
+literals. Because the value of the literal is carried in the type, a type-level
+decision can be made as to whether the conversion is valid.
+
+The integer types defined in the standard library permit conversion from integer
+literal types whose values are representable in the integer type. The
+floating-point types defined in the Carbon library permit conversion from
+integer and rational literal types whose values are between the minimum and
+maximum finite value representable in the floating-point type.
+
+### Defined Types
+
+The following types are defined in the Carbon prelude:
+
+-   An arbitrary-precision integer type.
+
+    ```
+    class BigInt;
+    ```
+
+-   A rational type, parameterized by a type used for its numerator and
+    denominator.
+
+    ```
+    class Rational(T:! Type);
+    ```
+
+    The exact constraints on `T` are not yet decided.
+
+-   A type representing integer literals.
+
+    ```
+    class IntLiteral(N:! BigInt);
+    ```
+
+-   A type representing floating-point literals.
+
+        ```
+        class FloatLiteral(X:! Rational(BigInt));
+        ```
+
+    All of these types are usable during compilation. `BigInt` supports the same
+    operations as `Int(n)`. `Rational(T)` supports the same operations as
+    `Float(n)`.
+
+The types `IntLiteral(n)` and `FloatLiteral(x)` also support primitive integer
+and floating-point operations such as arithmetic and comparison, but these
+operations are typically heterogeneous: for example, an addition between
+`IntLiteral(n)` and `IntLiteral(m)` produces a value of type
+`IntLiteral(n + m)`.
+
+### Integers
+
+Integer types can be either signed or unsigned, much like in C++. Signed
+integers are represented using 2's complement and notionally modeled as
+unbounded natural numbers. Signed overflow in either direction is an error.
+Specific sizes are available, for example: `i8`, `u16`, `i32`, and `u128`.
+
+There is an upper bound on the size of an integer, most likely initially set to
+128 bits due to LLVM limitations.
+
+### Floats
+
+Floating point types are based on the binary floating point formats provided by
+IEEE-754. `f16`, `f32`, `f64` and, if available, `f128` correspond exactly to
+those sized IEEE-754 formats, and have the semantics defined by IEEE-754.
+
+### Implicit conversions
+
+`IntLiteral(n)` converts to any sufficiently large integer type, as if by:
+
+```
+impl [template N:! BigInt, template M:! BigInt]
+    IntLiteral(N) as ImplicitAs(Int(M))
+    if N >= Int(M).MinValue as BigInt and N <= Int(M).MaxValue as BigInt {
+  ...
+}
+impl [template N:! BigInt, template M:! BigInt]
+    IntLiteral(N) as ImplicitAs(Unsigned(M))
+    if N >= Int(M).MinValue as BigInt and N <= Int(M).MaxValue as BigInt {
```

Comment:
See https://github.com/carbon-language/carbon-lang/issues/1998#issuecomment-1212644291 (wouldn't be bad to add that to the references) these types are now spelled `Carbon.Int(M)` and `Carbon.UInt(M)`.

```suggestion
    IntLiteral(N) as ImplicitAs(Int(M))
    if N >= Carbon.Int(M).MinValue as BigInt
    and N <= Carbon.Int(M).MaxValue as BigInt {
  ...
}
impl [template N:! BigInt, template M:! BigInt]
    IntLiteral(N) as ImplicitAs(Carbon.UInt(M))
    if N >= Carbon.UInt(M).MinValue as BigInt
    and N <= Carbon.UInt(M).MaxValue as BigInt {
```